x-pack/plugins/enterprise_search/public/applications/shared/role_mapping/users_table.tsx

```diff
@@ -0,0 +1,145 @@
+  ALL_LABEL,
+  FILTER_USERS_LABEL,
+  NO_USERS_LABEL,
+} from './constants';
+
+import { UsersAndRolesRowActions } from './';
+
+interface AccessItem {
+  name: string;
+}
+
+interface SharedUser extends SingleUserRoleMapping<ASRoleMapping | WSRoleMapping> {
+  accessItems: AccessItem[];
+  username: string;
+  email: string | null;
+  roleType: string;
+  id: string;
+}
+
+interface SharedRoleMapping extends ASRoleMapping, WSRoleMapping {
+  accessItems: AccessItem[];
+}
+
+interface Props {
+  accessItemKey: 'groups' | 'engines';
+  singleUserRoleMappings: Array<SingleUserRoleMapping<ASRoleMapping | WSRoleMapping>>;
+  initializeSingleUserRoleMapping(roleId: string): string;
+  handleDeleteMapping(roleId: string): string;
+}
+
+const noItemsPlaceholder = <EuiTextColor color="subdued">&mdash;</EuiTextColor>;
+const invitationBadge = <EuiBadge color="hollow">{INVITATION_PENDING_LABEL}</EuiBadge>;
+
+export const UsersTable: React.FC<Props> = ({
+  accessItemKey,
+  singleUserRoleMappings,
+  initializeSingleUserRoleMapping,
+  handleDeleteMapping,
+}) => {
+  // 'accessItems' is needed because App Search has `engines` and Workplace Search has `groups`.
+  const users = ((singleUserRoleMappings as SharedUser[]).map((user) => ({
+    username: user.elasticsearchUser.username,
+    email: user.elasticsearchUser.email,
+    roleType: user.roleMapping.roleType,
+    id: user.roleMapping.id,
+    accessItems: (user.roleMapping as SharedRoleMapping)[accessItemKey],
+    invitation: user.invitation,
+  })) as unknown) as Array<Omit<SharedUser, 'elasticsearchUser | roleMapping'>>;
+
+  const columns: Array<EuiBasicTableColumn<SharedUser>> = [
+    {
+      field: 'username',
+      name: 'Username',
+      render: (_, { username }: SharedUser) => username,
+    },
+    {
+      field: 'email',
+      name: 'Email',
+      render: (_, { email, invitation }: SharedUser) => {
+        if (!email) return noItemsPlaceholder;
+        return (
+          <div data-test-subj="EmailDisplayValue">
+            {email} {invitation && invitationBadge}
+          </div>
+        );
+      },
+    },
+    {
+      field: 'roleType',
+      name: 'Role',
+      render: (_, user: SharedUser) => user.roleType,
+    },
+    {
+      field: 'accessItems',
+      name: upperFirst(accessItemKey),
+      render: (_, { accessItems }: SharedUser) => {
+        // Design calls for showing the first 2 items followed by a +x after those 2.
+        // ['foo', 'bar', 'baz'] would display as: "foo, bar + 1"
+        const numItems = accessItems.length;
+        if (numItems === 0) return <span data-test-subj="AllItems">{ALL_LABEL}</span>;
+        const additionalItems = numItems > 2 ? ` + ${numItems - 2}` : '';
+        const names = accessItems.map((item) => item.name);
+        return (
+          <span data-test-subj="AccessItems">{names.slice(0, 2).join(', ') + additionalItems}</span>
+        );
+      },
+    },
+    {
+      field: 'id',
+      name: '',
```

Comment:
This needs an `Actions` table column heading for screen reader accessibility.

 TBH, generally I'd suggest using EUI's baked in `actions` column API (https://elastic.github.io/eui/#/tabular-content/tables#adding-actions-to-table), where you just pass in an array of action objs. It renders the table heading for you

I realize this would require a decent amount of refactoring (prob away from a React component and to regular fn), so I'm fine punting on this until after this PR if needed.

Reply:
Added to Google doc for tech debt after 7.14 FF